File: impl/csharp/test/ConformanceTest.cs
// Conformance test for the N-PAMP C# reference (draft-bubblefish-npamp-00).
// Exits 0 on success, 1 if any assertion fails. Mirrors the Go/Rust/Java suites.
using System;
using System.Text;

namespace Sh.Bubblefish.Npamp;

public static class ConformanceTest
{
    private static int _failures;

    private static void Check(string name, bool ok)
    {
        if (ok)
        {
            Console.WriteLine("ok   - " + name);
        }
        else
        {
            Console.WriteLine("FAIL - " + name);
            _failures++;
        }
    }

    private static byte[] Ramp(int start, int count)
    {
        byte[] b = new byte[count];
        for (int i = 0; i < count; i++)
        {
            b[i] = (byte)(start + i);
        }
        return b;
    }

    // --- cross-language vector reproduction (values from the Go reference) ---

    private static void VecHeader()
    {
        var f = new Npamp.Frame(Npamp.FramePing, Npamp.ChanControl);
        Check("vec_header", Npamp.ToHex(f.Marshal()) ==
            "4e50414d20000100000000000000000000000000000d880c250000000000000000000000");
    }

    private static void VecNonce()
    {
        Check("vec_nonce", Npamp.ToHex(Npamp.DeriveNonce(Ramp(0x01, 12), 0x0102030405060708UL)) ==
            "010203040404040c0c0c0c04");
    }

    private static void VecAead()
    {
        byte[] aad = new Npamp.Frame(Npamp.FramePing, Npamp.ChanControl).HeaderPrefix(11);
        byte[] sealedBytes = Npamp.SealAes256Gcm(Ramp(0x00, 32), Ramp(0x10, 12), 7UL, aad,
            Encoding.ASCII.GetBytes("hello world"));
        Check("vec_aead", Npamp.ToHex(sealedBytes) ==
            "3fe8b79f95b5697926b3395429c2c2466999c652f9346aeebb30bf");
    }

    private static void VecTrafficKey()
    {
        byte[] master = new byte[48];
        Array.Fill(master, (byte)0x2A);
        byte[] ts = Npamp.DeriveTrafficSecret(master, 0, 0UL, Npamp.AeadAes256Gcm, Npamp.ChanControl, false);
        Check("vec_traffic_key", Npamp.ToHex(Npamp.DeriveKeyIv(ts, false).Key) ==
            "79372e2fb7f92d63e3a68099ff72514f310ebf6773deb0fa7ef45d013c652dcc");
    }

    // --- property tests (mirror the Go/Rust/Java suites) ---

    private static void Roundtrip()
    {
        var f = new Npamp.Frame(0x0100, Npamp.ChanMemory, 42UL, Npamp.FlagEnc, 0,
            Encoding.ASCII.GetBytes("payload"));
        var g = Npamp.Frame.Unmarshal(f.Marshal());
        Check("roundtrip", g.Flags == Npamp.FlagEnc && g.Ftype == 0x0100 && g.Channel == Npamp.ChanMemory
            && g.Seq == 42UL && Encoding.ASCII.GetString(g.Payload) == "payload");
    }

    private static void CrcValidatedFirst()
    {
        byte[] buf = new Npamp.Frame(Npamp.FramePing, Npamp.ChanControl).Marshal();
        buf[5] ^= 0xFF; // corrupt frame-type byte; CRC must reject before any field is trusted
        bool rejected = false;
        try { Npamp.Frame.Unmarshal(buf); }
        catch (FrameException e) { rejected = e.Message == "bad crc"; }
        Check("crc_validated_first", rejected);
    }

    private static void ReservedMustBeZero()
    {
        byte[] buf = new Npamp.Frame(Npamp.FramePing, Npamp.ChanControl).Marshal();
        buf[30] = 1; // a reserved octet
        bool rejected = false;
        try { Npamp.Frame.Unmarshal(buf); }
        catch (FrameException e) { rejected = e.Message == "bad crc" || e.Message == "reserved nonzero"; }
        Check("reserved_must_be_zero", rejected);
    }

    private static void AeadTamperFails()
    {
        byte[] key = new byte[32];
        byte[] iv = Ramp(0x10, 12);
        byte[] aad = new Npamp.Frame(Npamp.FramePing, Npamp.ChanControl).HeaderPrefix(5);
        byte[] sealedBytes = Npamp.SealAes256Gcm(key, iv, 7UL, aad, Encoding.ASCII.GetBytes("hello"));
        bool openOk = Encoding.ASCII.GetString(Npamp.OpenAes256Gcm(key, iv, 7UL, aad, sealedBytes)) == "hello";
        aad[5] ^= 1;
        bool tamperRejected = false;
        try { Npamp.OpenAes256Gcm(key, iv, 7UL, aad, sealedBytes); }
        catch (Exception) { tamperRejected = true; }
        Check("aead_tamper_fails", openOk && tamperRejected);
    }

    private static void HkdfPrefixProtocolSpecific()
    {
        Check("hkdf_prefix_protocol_specific", Npamp.LabelPrefix == "n-pamp " && Npamp.LabelPrefix != "tls13 ");
    }

    public static int Main()
    {
        VecHeader();
        VecNonce();
        VecAead();
        VecTrafficKey();
        Roundtrip();
        CrcValidatedFirst();
        ReservedMustBeZero();
        AeadTamperFails();
        HkdfPrefixProtocolSpecific();
        Console.WriteLine(_failures == 0 ? "ALL PASS (9/9)" : ("FAILURES: " + _failures));
        return _failures == 0 ? 0 : 1;
    }
}

File: impl/php/test/ConformanceTest.php
<?php

/**
 * Conformance test for the N-PAMP PHP reference (draft-bubblefish-npamp-00).
 *
 * Reproduces the four cross-language golden vectors plus five property tests,
 * mirroring the Go/Rust/Java/Python/C# suites. Prints one line per check and
 * exits 0 on success, 1 if any check fails.
 *
 * Run from impl/php:  php test/ConformanceTest.php
 */

declare(strict_types=1);

require __DIR__ . '/../src/Npamp.php';

use Sh\Bubblefish\Npamp\Frame;
use Sh\Bubblefish\Npamp\FrameException;
use Sh\Bubblefish\Npamp\Npamp;

$failures = 0;

function check(string $name, bool $ok): void
{
    global $failures;
    if ($ok) {
        echo "ok   - {$name}\n";
    } else {
        echo "FAIL - {$name}\n";
        $failures++;
    }
}

function ramp(int $start, int $count): string
{
    $out = '';
    for ($i = 0; $i < $count; $i++) {
        $out .= chr(($start + $i) & 0xFF);
    }
    return $out;
}

// --- cross-language vector reproduction (values from the Go reference) ---

$f = new Frame(Npamp::FRAME_PING, Npamp::CHAN_CONTROL, 0);
check('vec_header', Npamp::toHex($f->marshal())
    === '4e50414d20000100000000000000000000000000000d880c250000000000000000000000');

check('vec_nonce', Npamp::toHex(Npamp::deriveNonce(ramp(0x01, 12), 0x0102030405060708))
    === '010203040404040c0c0c0c04');

$key = ramp(0x00, 32);
$iv = ramp(0x10, 12);
$aad = (new Frame(Npamp::FRAME_PING, Npamp::CHAN_CONTROL))->headerPrefix(11);
$sealed = Npamp::sealAes256Gcm($key, $iv, 7, $aad, 'hello world');
check('vec_aead', Npamp::toHex($sealed)
    === '3fe8b79f95b5697926b3395429c2c2466999c652f9346aeebb30bf');

$master = str_repeat(chr(0x2A), 48);
$ts = Npamp::deriveTrafficSecret($master, 0, 0, Npamp::AEAD_AES256_GCM, Npamp::CHAN_CONTROL, false);
[$tk] = Npamp::deriveKeyIv($ts, false);
check('vec_traffic_key', Npamp::toHex($tk)
    === '79372e2fb7f92d63e3a68099ff72514f310ebf6773deb0fa7ef45d013c652dcc');

// --- property tests (mirror the Go/Rust/Java suites) ---

$f = new Frame(0x0100, Npamp::CHAN_MEMORY, 42, Npamp::FLAG_ENC, 0, 'payload');
$g = Frame::unmarshal($f->marshal());
check('roundtrip', $g->flags === Npamp::FLAG_ENC && $g->ftype === 0x0100
    && $g->channel === Npamp::CHAN_MEMORY && $g->seq === 42 && $g->payload === 'payload');

// Corrupt the frame-type byte; CRC must reject before any field is trusted.
$buf = (new Frame(Npamp::FRAME_PING, Npamp::CHAN_CONTROL))->marshal();
$buf[5] = chr(ord($buf[5]) ^ 0xFF);
$rejected = false;
try {
    Frame::unmarshal($buf);
} catch (FrameException $e) {
    $rejected = ($e->getMessage() === 'bad crc');
}
check('crc_validated_first', $rejected);

// Set a reserved octet (byte 30). CRC covers only the 21-byte prefix, so it still
// passes; the reserved-zero check must then reject the frame.
$buf = (new Frame(Npamp::FRAME_PING, Npamp::CHAN_CONTROL))->marshal();
$buf[30] = chr(1);
$rejected = false;
try {
    Frame::unmarshal($buf);
} catch (FrameException $e) {
    $rejected = ($e->getMessage() === 'bad crc' || $e->getMessage() === 'reserved nonzero');
}
check('reserved_must_be_zero', $rejected);

// Seal, confirm open round-trips, then flip an AAD byte: open must reject.
$key = str_repeat("\0", 32);
$iv = ramp(0x10, 12);
$aad = (new Frame(Npamp::FRAME_PING, Npamp::CHAN_CONTROL))->headerPrefix(5);
$sealed = Npamp::sealAes256Gcm($key, $iv, 7, $aad, 'hello');
$openOk = (Npamp::openAes256Gcm($key, $iv, 7, $aad, $sealed) === 'hello');
$aad[5] = chr(ord($aad[5]) ^ 1);
$tamperRejected = false;
try {
    Npamp::openAes256Gcm($key, $iv, 7, $aad, $sealed);
} catch (\Throwable $e) {
    $tamperRejected = true;
}
check('aead_tamper_fails', $openOk && $tamperRejected);

check('hkdf_prefix_protocol_specific',
    Npamp::LABEL_PREFIX === 'n-pamp ' && Npamp::LABEL_PREFIX !== 'tls13 ');

echo $failures === 0 ? "ALL PASS (9/9)\n" : "FAILURES: {$failures}\n";
exit($failures === 0 ? 0 : 1);
